Add `:log10`-`colorbar_scale` support in GR

lmh91 opened the pull request:
Relates to https://github.com/JuliaPlots/Plots.jl/issues/3174
I added support for log10 scale of the colormap in heatmaps in GR. Please have a look. 

Notes:
* It is not possible that `z` contains negative values.
* In the non-uniform case, "pixels" with `z < clims[1]` are plotted white and not transparent as it seems that this is not yet supported by GR.jl (though I am not sure about that).

Some example: 

```julia
using Plots

x = 0:0.2:1;
# y = [0.0, 0.2, 0.4, 0.6]; # also works for evenly spaced heatmaps
y = [0.0, 0.15, 0.45, 0.6]; 
z = Array{Float64,2}(undef, length(y)-1, length(x)-1);
map(i -> z[i] = 10.0^(i-5), eachindex(z));

plot(
    plot(x, y, z, st=:heatmap, colorbar_scale = :identity),
    plot(x, y, z, st=:heatmap, colorbar_scale = :log10),
    plot(x, y, z, st=:heatmap, colorbar_scale = :log10, clims = (z[4], z[end-3])),
    layout = (3, 1), size = (800, 900)
)

```
![clog_heatmaps_gr](https://user-images.githubusercontent.com/16963906/139579730-c67a1721-23ee-4c6e-b71d-cd4946f0807c.png)

BeastyBlacksmith commented:
> "pixels" with z < clims[1] are plotted white and not transparent

can you make that picking up the background color?

lmh91 commented:
> > "pixels" with z < clims[1] are plotted white and not transparent
> 
> can you make that picking up the background color?

I changed it. 

I also rethought this and changed the behavior with `clims`:
Voxels outside the `clims` are plotted with the color corresponding to `cmin` and `cmax`. 
Only voxels where `z is NaN` are plotted transparent or with the background color. 

I think that gives the user more control. What do you think?

Here the updated and extended example:

![clog_heatmaps_gr](https://user-images.githubusercontent.com/16963906/139801772-8bf4503a-e0bb-4097-b50f-05e8e20b8de2.png)

lmh91 commented:
By the way: I take the `log10` of the data (`z`). 
In principle, I think that is not the right approach (performance wise)
One should make a logarithmic scaled colormap. 
But I don't understand how/where `fillgrad = _as_gradient(series[:fillcolor])` is defined
or actually `series[:fillcolor]` and how it is connected to the colorbar in `gr_draw_colorbar`.

BeastyBlacksmith commented:
`_as_gradient` is defined here: https://github.com/JuliaPlots/Plots.jl/blob/2e67e843611b9a13c568e7343b119c036ce9bd30/src/utils.jl#L206

and `cgrad` is defined in PlotUtils.jl

oschulz commented:
Thanks a lot for this @lmh91 !

lmh91 commented:
But where does `series[:fillcolor]` get initialized?
It is already a colormap (`::PlotUtils.ContinuousColorGradient`) inside `gr_draw_heatmap`. 

I would need to change `series[:fillcolor]` depending on `colorbar_scale`.

BeastyBlacksmith commented:
There is `get_fillcolor` which is hard to find: https://github.com/JuliaPlots/Plots.jl/blob/2e67e843611b9a13c568e7343b119c036ce9bd30/src/utils.jl#L497

and a similar construct in `_update_series_attributes!`: https://github.com/JuliaPlots/Plots.jl/blob/2e67e843611b9a13c568e7343b119c036ce9bd30/src/args.jl#L2162

lmh91 commented:
Thanks. With your help I was able to track it back to 

https://github.com/JuliaPlots/Plots.jl/blob/fdbd2c397c28bd29bfc3bf317ff961cca2185fe7/src/args.jl#L2070

However, it turned out I still have to take the logarithm to map `z` in the `[0,1]`.

Can we take it like it is?

BeastyBlacksmith reviewed (approved):
> Pixels outside the clims are plotted with the color corresponding to cmin and cmax.

I can't find references right now, but I remember people asking for having the ability to specify a color outside of the colormap for this. But this can also come in another PR.

I am fine with merging as is.

lmh91 commented:
Alright, then I would leave it like it is for now.

zhiyuanzhai commented:
Does this only work for `heatmap`? It doesn't seem to work for `contour` in my environment...
